### ISSUE TITLE
New source initialization broken

### ISSUE THREAD

**Issue by pmelchior** [pmelchior/scarlet#136]
The code from our quickstart guide
```
sources = []
for src in catalog:
    try:
        new_source = scarlet.ExtendedSource(frame, (src['y'], src['x']), observation, bg_rms)
    except scarlet.SourceInitError:
        try:
            new_source = scarlet.PointSource(frame, (src['y'], src['x']), observation)
        except scarlet.SourceInitError:
            print("Could not initialze source at {0}".format((src['y'], src['x'])))
            continue
    sources.append(new_source)
```
yields this warning
```
/Users/pmelchior/code/scarlet/scarlet/measurement.py:84: RuntimeWarning: invalid value encountered in double_scalars
  first_moment_y = np.sum(indy*morph_view_weighted) / morph_view_weighted_sum
/Users/pmelchior/code/scarlet/scarlet/measurement.py:85: RuntimeWarning: invalid value encountered in double_scalars
  first_moment_x = np.sum(indx*morph_view_weighted) / morph_view_weighted_sum
/Users/pmelchior/code/scarlet/scarlet/operator.py:206: RuntimeWarning: overflow encountered in long_scalars
  dy = int(2*(py-cy))
/Users/pmelchior/code/scarlet/scarlet/operator.py:207: RuntimeWarning: overflow encountered in long_scalars
  dx = int(2*(px-cx))
/Users/pmelchior/code/scarlet/scarlet/update.py:65: RuntimeWarning: invalid value encountered in true_divide
  component.morph[:] = component.morph / norm
```
which results in source seds to be 0 and morphs to be nan

**Comment by pmelchior** [comment]
The warning above is a red herring, the measurement module is not at fault. It is fed improper morphologies. I think the origin is that the sed is zero, after which the normalization is ill-defined and the morphs go to nan.

The problem arises for PointSource and ExtendedSource, but only if the observations have psf set.

**Comment by pmelchior** [comment]
Another false herring. The problem is caused by `update.monotonic` in `source.update`, it sets all morph pixels to zero. Again, only if PSF is set

**Comment by pmelchior** [comment]
`update.monotonic` does not care about PSFs. So, the problem is the initial morphology that is passed to the prox. The offending part is the new `Fourier.image` code and the slicing here https://github.com/pmelchior/scarlet/blob/master/scarlet/source.py#L373:L374

`bb` from this code is a 1-pixel wide (in horizontal) strip over the entire image. The slices are matched (full vertical, 1 pixel horizontal), but they don't have the correct shape or the correct location.

**Comment by pmelchior** [comment]
The problem looks it's related to how the PSFs are now set up. This is what we use in the quickstart tutorial
```
model_psf = scarlet.psf.gaussian(coords, y0, x0, 1, .9)
model_psf /= model_psf.sum()
```
That seems to be what the new version wants:
```
model_psf = scarlet.psf.generate_psf_image(scarlet.psf.gaussian, psfs.shape[1:])
```
However, the source initialization works now, but `blend.fit` raises an error when calling `model_ = self._convolve(model_)`.

How is the new PSF structure supposed to be created and called?

**Comment by pmelchior** [comment]
Even the updated `user_docs.ipynb`, which uses
```
model_psf = scarlet.psf.generate_psf_image(scarlet.psf.gaussian, psfs.shape[1:], sigma=.9, amplitude=1)[None]
```
works for source initialization, but breaks on `blend.fit`. How could this get merged to master?

**Comment by pmelchior** [comment]
For reference, here's the error trace:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<timed eval> in <module>

~/code/scarlet/scarlet/blend.py in fit(self, max_iter, e_rel, approximate_L)
     79         for step in range(max_iter):
     80             # Back propagate the gradients
---> 81             self._backward()
     82 
     83             # Calculate the Lipschitz constants,

~/code/scarlet/scarlet/blend.py in _backward(self)
    111         # This calculates the partial derivatives wrt
    112         # all the seds and morphologies
--> 113         gradients = self._grad(*parameters)
    114         sed_gradients = gradients[:self.K]
    115         morph_gradients = gradients[self.K:]

/anaconda3/lib/python3.7/site-packages/autograd/wrap_util.py in nary_f(*args, **kwargs)
     18             else:
     19                 x = tuple(args[i] for i in argnum)
---> 20             return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)
     21         return nary_f
     22     return nary_operator

/anaconda3/lib/python3.7/site-packages/autograd/differential_operators.py in grad(fun, x)
     22     arguments as `fun`, but returns the gradient instead. The function `fun`
     23     should be scalar-valued. The gradient has the same type as the argument."""
---> 24     vjp, ans = _make_vjp(fun, x)
     25     if not vspace(ans).size == 1:
     26         raise TypeError("Grad only applies to real scalar-output functions. "

/anaconda3/lib/python3.7/site-packages/autograd/core.py in make_vjp(fun, x)
      8 def make_vjp(fun, x):
      9     start_node = VJPNode.new_root(x)
---> 10     end_value, end_node =  trace(start_node, fun, x)
     11     if end_node is None:
     12         def vjp(g): return vspace(x).zeros()

/anaconda3/lib/python3.7/site-packages/autograd/tracer.py in trace(start_node, fun, x)
      8     with trace_stack.new_trace() as t:
      9         start_box = new_box(x, t, start_node)
---> 10         end_box = fun(start_box)
     11         if isbox(end_box) and end_box._trace == start_box._trace:
     12             return end_box._value, end_box._node

/anaconda3/lib/python3.7/site-packages/autograd/wrap_util.py in unary_f(x)
     13                 else:
     14                     subargs = subvals(args, zip(argnum, x))
---> 15                 return fun(*subargs, **kwargs)
     16             if isinstance(argnum, int):
     17                 x = args[argnum]

~/code/scarlet/scarlet/blend.py in _loss(self, *parameters)
    135         total_loss = 0
    136         for observation in self.observations:
--> 137             total_loss = total_loss + observation.get_loss(model)
    138         self.mse.append(total_loss._value)
    139         return total_loss

~/code/scarlet/scarlet/observation.py in get_loss(self, model)
    294         """
    295 
--> 296         model = self.render(model)
    297 
    298         return 0.5 * np.sum((self.weights * (model - self.images)) ** 2)

~/code/scarlet/scarlet/observation.py in render(self, model)
    275         model_ = model[self._band_slice, :, :]
    276         if self._diff_kernels is not None:
--> 277             model_ = self._convolve(model_)
    278 
    279         return model_

~/code/scarlet/scarlet/observation.py in _convolve(self, model)
    258         """Convolve the model in a single band
    259         """
--> 260         return fft.convolve(fft.Fourier(model, axes=(1, 2)), self._diff_kernels).image
    261 
    262     def render(self, model):

~/code/scarlet/scarlet/fft.py in convolve(image1, image2, padding, axes)
    312         to supress artifacts.
    313     """
--> 314     return _kspace_operation(image1, image2, padding, operator.mul, image1.shape)

~/code/scarlet/scarlet/fft.py in _kspace_operation(image1, image2, padding, op, shape)
    272         raise Exception(msg)
    273     fft_shape = _get_fft_shape(image1.image, image2.image, padding, image1.axes)
--> 274     convolved_fft = op(image1.fft(fft_shape), image2.fft(fft_shape))
    275     convolved = Fourier.from_fft(convolved_fft, fft_shape, shape, image1.axes)
    276     return convolved

~/code/scarlet/scarlet/fft.py in fft(self, fft_shape)
    199                 msg = "fft_shape self.axes must have the same number of dimensions, got {0}, {1}"
    200                 raise ValueError(msg.format(fft_shape, self.axes))
--> 201             image = _pad(self.image, fft_shape, self._axes)
    202             self._fft[fft_shape] = np.fft.rfftn(np.fft.ifftshift(image, self._axes), axes=self._axes)
    203         return self._fft[fft_shape]

~/code/scarlet/scarlet/fft.py in _pad(arr, newshape, axes)
     63             endind = dS - startind
     64             pad_width[axis] = (startind, endind)
---> 65     return np.pad(arr, pad_width, mode="constant")
     66 
     67 

/anaconda3/lib/python3.7/site-packages/autograd/tracer.py in f_wrapped(*args, **kwargs)
     43             argnums = tuple(argnum    for argnum, _   in boxed_args)
     44             ans = f_wrapped(*argvals, **kwargs)
---> 45             node = node_constructor(ans, f_wrapped, argvals, kwargs, argnums, parents)
     46             return new_box(ans, trace, node)
     47         else:

/anaconda3/lib/python3.7/site-packages/autograd/core.py in __init__(self, value, fun, args, kwargs, parent_argnums, parents)
     28     def __init__(self, value, fun, args, kwargs, parent_argnums, parents):
     29         self.parents = parents
---> 30         self.vjp = primitive_vjps[fun](parent_argnums, value, args, kwargs)
     31 
     32     def initialize_root(self, value):

KeyError: <function primitive.<locals>.f_wrapped at 0xa13ff0d90>
```

**Comment by herjy** [comment]
Updadte autograd. Only the latest version of autograd.numpy has pad.

**Comment by fred3m** [comment]
Really, this is an autograd thing again? @pmelchior should have the right autograd version unless this is a fresh install.

**Comment by pmelchior** [comment]
OK, then you need to document that an upgrade is needed. The canonical way of doing that is to require a specific version as install dependency in `setup.py`.

I will try to confirm that it's `autograd`

**Comment by pmelchior** [comment]
Confirmed. I will make a quick commit to update setup.py

**Comment by fred3m** [comment]
Agreed. We noted that when @herjy  had an issue, but have not pushed a branch with the updated docs.

Part of the reason that we didn't specify a version of `autograd` is because the package is maintained but no longer developed, so while `pad` support was added last year a release of autograd that contained `pad` was only just added recently. In general I think that the docs should recommend that users use the latest `master` from `autograd` (which is what we use in the stack and I thought that is what we have all been using, since there was no working release with FFTs when we switched to autograd). In fact, I'm surprised that you have a version that even supports `np.fft`, since I didn't think it was in the previous release.
